lib/xnvme_be_macos_driverkit_async.c

```diff
@@ -0,0 +1,219 @@
+	if (res != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectMapMemory64(); "
+			    "ret(0x%08x), '%s'",
+			    res, mach_error_string(res));
+		return -EIO;
+	}
+	q->cq_queue = (RingQueue *)addr;
+
+	return 0;
+}
+
+int
+xnvme_be_macos_driverkit_queue_term(struct xnvme_queue *_queue)
+{
+	struct xnvme_queue_macos_driverkit *q = (struct xnvme_queue_macos_driverkit *)_queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)q->base.dev->be.state;
+	int ret;
+
+	ret = _xnvme_be_driverkit_delete_ioqpair(state, q->id);
+	if (ret) {
+		XNVME_DEBUG("FAILED: _xnvme_be_driverkit_delete_ioqpair()");
+		return ret;
+	}
+
+	q->cq_queue = NULL;
+	q->sq_queue = NULL;
+	return ret;
+}
+
+int
+xnvme_be_macos_driverkit_queue_poke(struct xnvme_queue *queue, uint32_t XNVME_UNUSED(max))
+{
+	struct xnvme_queue_macos_driverkit *q = (struct xnvme_queue_macos_driverkit *)queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)queue->base.dev->be.state;
+	uint64_t qid[2] = {q->id, q->id}; // Submission queue, completion queue
+	int reaped = 0;
+	int queue_ret;
+	NvmeSubmitCmd nvme_cmd;
+	kern_return_t ret;
+	struct xnvme_cmd_ctx *ctx;
+
+	ret = IOConnectCallScalarMethod(state->device_connection, NVME_POKE, qid, 2, NULL, 0);
+	if (ret != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectCallScalarMethod(); ret(0x%x), msg(%s)", ret,
+			    mach_error_string(ret));
+		return -EIO;
+	}
+	while ((queue_ret = queue_dequeue(q->cq_queue, &nvme_cmd)) == 0) {
+		ctx = (struct xnvme_cmd_ctx *)nvme_cmd.backend_opaque;
+		memcpy(&ctx->cpl, nvme_cmd.cpl, sizeof(struct xnvme_spec_cpl));
+		ctx->async.cb(ctx, ctx->async.cb_arg);
+
+		reaped += 1;
+		queue->base.outstanding -= 1;
+	}
+
+	if (queue_ret == QUEUE_ERROR) {
+		XNVME_DEBUG("FAILED: queue_ret == QUEUE_ERROR");
+		return -EIO;
+	}
+
+	return reaped;
+}
+
+int
+xnvme_be_macos_driverkit_async_cmd_io(struct xnvme_cmd_ctx *ctx, void *dbuf, size_t dbuf_nbytes,
+				      void *mbuf, size_t mbuf_nbytes)
+{
+	struct xnvme_queue_macos_driverkit *q =
+		(struct xnvme_queue_macos_driverkit *)ctx->async.queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)ctx->dev->be.state;
+	NvmeSubmitCmd cmd = {0};
+	int queue_ret;
+	struct buffer *buf;
+	uint64_t _buf_base_offset;
+
+	cmd.queue_id = state->qid_sync;
+
+	switch (ctx->cmd.common.opcode) {
+	case XNVME_SPEC_NVM_OPC_READ:
+		break;
+
+	case XNVME_SPEC_FS_OPC_READ:
+		ctx->cmd.nvm.slba = ctx->cmd.nvm.slba >> ctx->dev->geo.ssw;
+		ctx->cmd.common.opcode = XNVME_SPEC_NVM_OPC_READ;
+		break;
+
+	case XNVME_SPEC_NVM_OPC_WRITE:
+		break;
+
+	case XNVME_SPEC_FS_OPC_WRITE:
+		ctx->cmd.nvm.slba = ctx->cmd.nvm.slba >> ctx->dev->geo.ssw;
+		ctx->cmd.common.opcode = XNVME_SPEC_NVM_OPC_WRITE;
+		break;
+	}
+
+	if (dbuf) {
+		buf = buffer_find(state->buffers, (uint64_t)dbuf);
+		if (!buf) {
```

Comment:
Should have ``XNVME_DEBUG("FAILED: ...")`` on the error-path here.

Reply:
Done.